http/tokenV2/middleware.go

```diff
@@ -0,0 +1,382 @@
+	}
+}
+
+// credentialIsSecure returns true,nil if a credential meets the minimum security
+// standards. This can cover things like sufficiently secure signing algorithms,
+// key size, etc.
+//
+// WARNING: A credential passing this check is not yet considered properly signed
+// or having valid essential claims such as NotBefore, Expiration, etc.
+func credentialIsSecure(credential string) error {
+	// Ignore very long credentials
+	if len(credential) > MaximumCredentialLength {
+		return fmt.Errorf("credential is too long")
+	}
+
+	// Parse the credential as a JWS (JSON Web Signature) containing a message. This works
+	// because a JWT (JSON Web Token) is built on a JWS where the claims are a signed message.
+	message, err := jws.ParseString(credential)
+	if err != nil {
+		return fmt.Errorf("cannot parse credential: jwk.ParseString: %w", err)
+	}
+
+	// Inspect the signatures in the message
+	secureSignatureCount := 0
+	for _, signature := range message.Signatures() {
+		// Reject credentials signed with insecure algorithms
+		algorithm := signature.ProtectedHeaders().Algorithm()
+		if !acceptableSignatureAlgorithm(algorithm) {
+			return fmt.Errorf("signing algorithm %v is not permitted", algorithm)
+		}
+
+		// Keep track of how many secure signatures are found
+		secureSignatureCount++
+	}
+
+	// Accept messages containing secure signatures
+	if secureSignatureCount > 0 {
+		return nil
+	}
+
+	// By default this method rejects messages
+	return fmt.Errorf("no signatures found")
+}
+
+// mandatoryJWTFields returns the mandatory fields of the JWT, and is effectively a constant
+func mandatoryJWTFields() []string {
+	return []string{jwt.JwtIDKey, jwt.IssuedAtKey, jwt.ExpirationKey, jwt.NotBeforeKey, jwt.AudienceKey, jwt.IssuerKey, jwt.SubjectKey}
+}
+
+// tokenJTI returns the JWT's JTI as a string
+func tokenJTI(token jwt.Token) string {
+	// Retrieve the JwtID from the token, but it comes out as an interface{}
+	if jtiIface, ok := token.Get(jwt.JwtIDKey); ok {
+		// Convert the interface{} to a string so the typed value can be returned
+		if jtiStr, ok := jtiIface.(string); ok {
+			return jtiStr
+		}
+	}
+
+	// Simply return an empty string if either the jti field wasn't present or it wasn't a string
+	return ""
+}
+
+// bestPracticesCheck ensures tokens are crafted in a sensible way, ensuring that even a valid signer must
+// conform to certain security controls such as not creating long lived credentials etc.
+//
+// WARNING: This check does not verify signatures or even the essential claims such as NotBefore, Expiration, etc.
+func bestPracticesCheck(token jwt.Token) error {
+	// Ensure the mandatory fields are present
+	for _, field := range mandatoryJWTFields() {
+		if _, ok := token.Get(field); !ok {
+			return fmt.Errorf("missing field: %v", field)
+		}
+	}
+
+	// Ensure JTI is a UUID
+	jti := tokenJTI(token)
+	if _, err := uuid.Parse(jti); err != nil {
```

Comment:
why require it if we don't do anything with it?

Reply:
It apparently serves to add randomness to the JWT. Wout asked about it, so I added it. It seems a bit "extra", but also not a bad idea